zone/spells.cpp

```diff
@@ -3211,29 +3205,55 @@ int Mob::CheckStackConflict(uint16 spellid1, int caster_level1, uint16 spellid2,
 // spells 1-50: no restrictons
 // 51-65: SpellLevel/2+15
 // 66+ Group Spells 62, Single Target 61
-bool Mob::CheckSpellLevelRestriction(uint16 spell_id)
+bool Mob::CheckSpellLevelRestriction(Mob *caster, uint16 spell_id)
 {
-	return true;
-}
+	bool check_for_restrictions = false;
+	bool can_cast = true;
 
-bool Client::CheckSpellLevelRestriction(uint16 spell_id)
-{
-	int SpellLevel = GetMinLevel(spell_id);
+	if (!caster) {
+		LogSpells("CheckSpellLevelRestriction: No caster");
+		return false;
+	}
 
-	// Only check for beneficial buffs
-	if (IsBuffSpell(spell_id) && IsBeneficialSpell(spell_id)) {
-		if (SpellLevel > 65) {
-			if (IsGroupSpell(spell_id) && GetLevel() < 62)
-				return false;
-			else if (GetLevel() < 61)
-				return false;
-		} else if (SpellLevel > 50) { // 51-65
-			if (GetLevel() < (SpellLevel / 2 + 15))
-				return false;
+	// NON GM clients might be restricted by rule setting
+	if (caster->IsClient()) {
+		if (RuleB(Spells, BuffLevelRestrictions) && !caster->CastToClient()->GetGM()) {
+			check_for_restrictions = true;
 		}
 	}
+	// NPCS might be restricted by rule setting
+	else if (RuleB(Spells, NPCBuffLevelRestrictions)) {
+		check_for_restrictions = true;
+	}
 
-	return true;
+	if (check_for_restrictions) {
+		int spell_level = GetMinLevel(spell_id);
+
+		// Only check for beneficial buffs
+		if (IsBuffSpell(spell_id) && IsBeneficialSpell(spell_id)) {
+			if (spell_level > 65) {
```

Comment:
Can we use 65 as a constant, magic number

Reply:
Same with the others below

Reply:
At the very least somewhere in `eq_constants.h`

Reply:
I'm happy to do this, but any names I come up with make the code less easy to understand IMO.  Like 65 is what, UPPER_BENEFICIAL_SPELL_LIMIT_THRESHOLD maybe?  And then 61 and 62 and 50?  I could use similar names but it feels like they make it less readable?